benchmarks/src/jmh/java/io/grpc/benchmarks/netty/AbstractBenchmark.java

```diff
@@ -0,0 +1,316 @@
+package io.grpc.benchmarks.netty;
+
+import com.google.common.util.concurrent.MoreExecutors;
+
+import io.grpc.ChannelImpl;
+import io.grpc.DeferredInputStream;
+import io.grpc.Marshaller;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.MethodType;
+import io.grpc.ServerCall;
+import io.grpc.ServerCallHandler;
+import io.grpc.ServerImpl;
+import io.grpc.ServerServiceDefinition;
+import io.grpc.Status;
+import io.grpc.stub.Calls;
+import io.grpc.stub.StreamObserver;
+import io.grpc.transport.netty.NegotiationType;
+import io.grpc.transport.netty.NettyChannelBuilder;
+import io.grpc.transport.netty.NettyServerBuilder;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.EmptyByteBuf;
+import io.netty.buffer.PooledByteBufAllocator;
+import io.netty.channel.local.LocalAddress;
+import io.netty.channel.local.LocalChannel;
+import io.netty.channel.local.LocalServerChannel;
+import io.netty.channel.nio.NioEventLoopGroup;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.net.SocketAddress;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+
+import javax.annotation.Nullable;
+
+/**
+ * Abstract base class for Netty end-to-end benchmarks.
+ */
+public abstract class AbstractBenchmark {
+
+  /**
+   * Standard payload sizes.
+   */
+  public static enum PayloadSize {
+    SMALL(10), MEDIUM(1024), LARGE(65536), JUMBO(16777216);
+
+    final int bytes;
+    PayloadSize(int bytes) {
+      this.bytes = bytes;
+    }
+
+    public int bytes() {
+      return bytes;
+    }
+  }
+
+  /**
+   * Standard flow-control window sizes.
+   */
+  public static enum FlowWindowSize {
+    SMALL(16384), MEDIUM(65536), LARGE(1048576), JUMBO(16777216);
+
+    final int bytes;
+    FlowWindowSize(int bytes) {
+      this.bytes = bytes;
+    }
+
+    public int bytes() {
+      return bytes;
+    }
+  }
+
+  /**
+   * Executor types used by Channel & Server.
+   */
+  public static enum ExecutorType {
+    DEFAULT, DIRECT;
+  }
+
+  /**
+   * Support channel types.
+   */
+  public static enum ChannelType {
+    NIO, LOCAL;
+  }
+
+
+  protected ServerImpl server;
+  protected ByteBuf request;
+  protected ByteBuf response;
+  protected MethodDescriptor unaryMethod;
+  protected ChannelImpl[] channels;
+
+  public AbstractBenchmark() {
+  }
+
+  /**
+   * Initialize the environment for the executor.
+   */
+  public void setup(ExecutorType clientExecutor,
```

Comment:
It will but there are way too many permutations if I do this, JMH estimated 17 hours to run if all permutations were enabled.  Hence creating subclasses for useful permutations